### PR TITLE
fix: restart the daemon after a clean SIGTERM (manager-owned lifecycle)

### PR DESCRIPTION
**Pull request by murdore**
## Problem

Shooter ran as a macOS LaunchAgent but silently died and never came back — ~17h of zero push notifications. Root cause, traced through the code:

1. The LaunchAgent's `ExecStart` runs `shooter start` in the **foreground**. That parent process spawns `server.ts` as a child and forwards signals to it.
2. On any SIGTERM, the parent forwards it → the server runs graceful shutdown → `process.exit(0)`.
3. The parent propagates the clean child exit → `process.exit(0)`.
4. The plist used `KeepAlive = { SuccessfulExit: false }` → launchd restarts **only on a non-zero (crash) exit**, so a clean exit-0 is deliberately **not** restarted. No reboot → `RunAtLoad` never re-fires → dead until next login.

So **any clean SIGTERM permanently kills the daemon**. (The auto-updater was ruled out — it no-ops on npm-global installs, which have no `.git`.)

## Fix — service-manager-owned lifecycle

Let launchd/systemd own restart instead of encoding intent in the exit code.

- **`KeepAlive = true` / `Restart = always`** — restart on *any* exit (crash, kill, or an accidental clean SIGTERM all recover).
- **Deliberate stop removes the job from the manager** (`launchctl bootout` / `systemctl --user stop`) so it stays stopped. The unit file remains on disk, so it returns on next login or `shooter start`.
- **Recursion guard:** the generated unit carries `SHOOTER_MANAGED=1`. A manual `shooter start` delegates to the manager; the manager-spawned instance (marker present, or `ppid === 1` on macOS as a transition fallback) runs the server directly — so the unit's `shooter start` never delegates to itself.

## Changes

- **New `bin/lib/service-manager.cjs`** — pure deciders `resolveStartAction` / `resolveStopAction` and unit-file builders `buildLaunchdPlist` / `buildSystemdUnit`.
- **`bin/shooter.cjs`** — `enableLaunchAgent` / `enableSystemdUnit` use the builders; `startServer` delegates to the manager; `stopServer` boots out / `systemctl stop`; new `isSystemdManaging()` + `isSpawnedByManager()` helpers.
- **`tests/service-manager.test.cjs`** — 13 unit tests (added to the `test` script).

## Testing

- `pnpm test` — full suite green, incl. the 13 new tests.
- Generated plist validated with `plutil -lint` (OK).
- `shooter version` / `help` smoke-tested.

## Deployment note

This goes live on a machine only after republish → `npm i -g @juspay/shooter@latest` → **`shooter autostart on`** (regenerates the plist with the new `KeepAlive` + marker). The auto-updater will not carry it (it no-ops on npm-global installs).

## Summary by CodeRabbit

* **New Features**
  * Improved service start/stop handling on macOS and Linux so the app can use the platform’s service manager when available.
  * Added stronger auto-restart behavior for managed installs, helping the app recover more reliably after clean exits.
  * Updated generated startup configuration for both platforms to include a managed-run marker.

* **Bug Fixes**
  * Reduced restart loops when stopping a managed instance.
  * Improved handling of special characters and spaces in generated service configuration.

### CODE REVIEW

File: bin/lib/service-manager.cjs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+/**
+ * Build the Linux systemd user unit.
+ *
+ * Restart=always (not on-failure) so a clean exit also recovers; a deliberate
+ * `systemctl --user stop` keeps it stopped. The SHOOTER_MANAGED marker mirrors
+ * the launchd path.
+ */
+function buildSystemdUnit({ nodeBin, shooterBin, pkgRoot, shooterHome }) {
```
Comment (Tara-ag) [inline bin/lib/service-manager.cjs:123]: ⚠️ **MAJOR: Incorrect systemd quoting for paths with spaces**

The current implementation generates:
```
ExecStart="/path with spaces/node" "/path with spaces/shooter.cjs" start
```

systemd's parser treats `"` as a regular character when not at the start of a token. Per systemd.exec(5), the correct way to quote paths with spaces is either:

1. **Single quotes around the whole value** (systemd strips one pair of outer quotes):
```ini
ExecStart='/path with spaces/node' '/path with spaces/shooter.cjs' start
```

2. **Double quotes around the whole value with inner quotes escaped**:
```ini
ExecStart="\"/path with spaces/node\" \"/path with spaces/shooter.cjs\" start"
```

The current approach will cause the service to fail to start when node or shooter are installed in paths containing spaces (common with usernames like "John Doe").

**Suggested fix:**
```javascript
function buildSystemdUnit({ nodeBin, shooterBin, pkgRoot, shooterHome }) {
  // Use single quotes for systemd - it strips one layer of outer quotes
  return `[Unit]\nDescription=Shooter — Mobile dev notifications & remote terminal\nAfter=network.target\n\n[Service]\nType=simple\nExecStart='${nodeBin}' '${shooterBin}' start\nWorkingDirectory='${pkgRoot}'\nEnvironment="SHOOTER_HOME=${shooterHome}"\nEnvironment="${MANAGED_ENV}=1"\nRestart=always\nRestartSec=10\n\n[Install]\nWantedBy=default.target\n`;
}
```

Note: Single quotes in paths would need escaping as `'\''` but that's extremely rare compared to spaces.

File: bin/shooter.cjs [hunk 2]
```diff
@@ -1153,6 +1140,96 @@ function isLaunchdManaging() {
...
+    execFileSync('systemctl', ['--user', 'is-active', '--quiet', 'shooter.service'], {
+      stdio: 'ignore',
+      timeout: 5_000,
+    });
+    return true;
+  } catch {
+    return false;
+  }
+}
+
```
Comment (Tara-ag) [inline bin/shooter.cjs:1161]: 💡 **MINOR: ppid === 1 check may be unreliable on macOS**

The fallback `process.ppid === 1` check assumes launchd reparents GUI agents to PID 1. However:

1. On modern macOS versions, launchd (PID 1) may not be the direct parent after certain system events
2. If the shooter process is spawned via `sudo` or other wrappers, ppid won't be 1
3. This is a transition fallback for old plists without `SHOOTER_MANAGED` marker

Consider adding a comment explaining this is intentionally a "best-effort" fallback that will be removed in a future version once all users have regenerated their plists. Also, the hardcoded UID 501 fallback on line 1149 should be documented as a "common default" rather than a guaranteed value.

**Suggested documentation:**
```javascript
// Transition fallback (vX.Y.Z): old plists without SHOOTER_MANAGED marker.
// launchd typically reparents GUI agents to PID 1, but this is not guaranteed.
// Remove this fallback after N releases once marker-based detection is universal.
```

File: tests/service-manager.test.cjs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  assertTrue(
+    /<key>SHOOTER_MANAGED<\/key>\s*<string>1<\/string>/.test(plist),
+    'marker env distinguishes the manager-spawned instance from a CLI invocation'
+  );
+});
+
+runTest('buildLaunchdPlist: keeps RunAtLoad and the start ProgramArguments', () => {
+  const plist = buildLaunchdPlist(PLIST_ARGS);
+  assertTrue(/<key>RunAtLoad<\/key>\s*<true\/>/.test(plist), 'RunAtLoad retained');
+  assertTrue(plist.includes('<string>/pkg/bin/shooter.cjs</string>'), 'shooterBin in args');
```
Comment (Tara-ag) [inline tests/service-manager.test.cjs:211]: 💡 **SUGGESTION: Add test for paths with spaces in systemd unit**

The current test only validates the `Restart=always` setting and marker presence. Given the systemd quoting complexity, consider adding a test that verifies paths with spaces are handled correctly:

```javascript
runTest('buildSystemdUnit: handles paths with spaces correctly', () => {
  const unit = buildSystemdUnit({
    nodeBin: '/home/John Doe/.nvm/versions/node/v20/bin/node',
    shooterBin: '/home/John Doe/.local/bin/shooter.cjs',
    pkgRoot: '/home/John Doe/projects/shooter',
    shooterHome: '/home/John Doe/.shooter',
  });
  // Verify the unit can be parsed by systemd (basic sanity check)
  assertTrue(unit.includes("ExecStart='"), 'paths should be single-quoted for systemd');
  assertTrue(unit.includes("WorkingDirectory='"), 'working dir should be single-quoted');
});
```

This would catch the quoting issue mentioned in the review of `bin/lib/service-manager.cjs`.

File: bin/shooter.cjs [hunk 4]
```diff
@@ -1153,6 +1140,96 @@ function isLaunchdManaging() {
   }
```
Comment (Tara-ag) [inline bin/shooter.cjs:1146]: 💡 **SUGGESTION: Handle missing systemctl gracefully**

The `isSystemdManaging()` function will throw if `systemctl` is not in PATH (e.g., on minimal containers or non-systemd Linux distributions). While the try/catch handles this, the error is silently swallowed which could make debugging difficult.

Consider logging the error at debug/trace level for troubleshooting:

```javascript
function isSystemdManaging() {
  if (os.platform() !== 'linux') return false;
  try {
    execFileSync('systemctl', ['--user', 'is-active', '--quiet', 'shooter.service'], {
      stdio: 'ignore',
      timeout: 5_000,
    });
    return true;
  } catch (err) {
    // ENOENT = systemctl not installed, non-zero exit = not active
    if (err.code === 'ENOENT') {
      // systemctl not available - expected on non-systemd systems
      return false;
    }
    return false;
  }
}
```

This is minor since the current behavior is functionally correct, but explicit handling improves maintainability.

File: tests/service-manager.test.cjs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    'KeepAlive should be <true/> so any clean exit still restarts'
+  );
+  assertTrue(
+    !plist.includes('SuccessfulExit'),
+    'SuccessfulExit gating must be gone (that is the bug being fixed)'
+  );
+});
+
+runTest('buildLaunchdPlist: injects the SHOOTER_MANAGED=1 marker', () => {
+  const plist = buildLaunchdPlist(PLIST_ARGS);
```
Comment (Tara-ag) [inline tests/service-manager.test.cjs:201]: 💡 **SUGGESTION: Expand XML escaping test coverage**

The current test only verifies `&` escaping. Consider expanding to cover `<` and `>` which are also escaped by `xmlEscapeText`:

```javascript
runTest('buildLaunchdPlist: XML-escapes all special characters', () => {
  const plist = buildLaunchdPlist({ 
    ...PLIST_ARGS, 
    pkgRoot: '/path<with>special&chars' 
  });
  assertTrue(plist.includes('&lt;'), '< is escaped to &lt;');
  assertTrue(plist.includes('&gt;'), '> is escaped to &gt;');
  assertTrue(plist.includes('&amp;'), '& is escaped to &amp;');
  // Verify no raw special chars exist in string values
  assertTrue(!/<string>[^<]*[<>&][^<]*<\/string>/.test(plist), 'no raw XML special chars in strings');
});
```

This ensures the escaping logic remains complete if the implementation changes.

File: bin/shooter.cjs [hunk 6]
```diff
@@ -1153,6 +1140,96 @@ function isLaunchdManaging() {
...
+  // Transition fallback for an older, marker-less plist still on disk: launchd
+  // reparents its GUI agents to PID 1, so a launchd-spawned `shooter start` has
+  // ppid 1 while a shell-run one does not. Prevents delegate-recursion.
+  if (os.platform() === 'darwin' && process.ppid === 1) return true;
+  return false;
+}
+
+// Hand a manual `shooter start` off to launchd. If the job is already loaded it
+// is already running (KeepAlive); otherwise bootstrap it (e.g. after a stop).
+function delegateStartToLaunchd() {
```
Comment (Tara-ag) [inline bin/shooter.cjs:1179]: 💡 **SUGGESTION: Race condition in delegateStartToLaunchd**

There's a TOCTOU (time-of-check/time-of-use) race between `isLaunchdManaging()` and the `launchctl bootstrap` call. If the service transitions between states (e.g., crashes and restarts) between the check and the bootstrap, the error handling may be confusing.

While this is unlikely in practice, consider making the idempotency explicit:

```javascript
function delegateStartToLaunchd() {
  const uid = process.getuid ? process.getuid() : 501;
  if (isLaunchdManaging()) {
    console.log('Shooter is already running (launchd-managed).');
    console.log('Run "shooter stop" to stop, or "shooter status" for details.');
    return;
  }
  try {
    execFileSync('launchctl', ['bootstrap', `gui/${uid}`, LAUNCHD_PLIST], { stdio: 'ignore' });
    console.log('Shooter started (launchd-managed).');
    console.log('  Logs:   shooter logs');
  } catch (err) {
    // Check if it became managed between our check and bootstrap attempt
    if (isLaunchdManaging()) {
      console.log('Shooter is already running (launchd-managed).');
      return;
    }
    console.error(`Failed to start via launchd: ${err.message}`);
    console.error('Try: shooter autostart on');
    process.exit(1);
  }
}
```

This is a minor improvement for robustness, not a blocker.

File: docs/superpowers/specs/2026-06-27-launchd-restart-resilience-design.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+`<true/>` and `SHOOTER_MANAGED`. Matches the existing `tests/` (vitest) setup.
+
+## Out of scope (YAGNI)
+
+Auto-updater, guard, APNs transport, tunnel — none are the bug.
+
+## Deployment
+
+Lands in repo → publishes `@juspay/shooter`. Live on a machine only after
+republish → `npm i -g @juspay/shooter@latest` → **`shooter autostart on`**
```
Comment (Tara-ag) [inline docs/superpowers/specs/2026-06-27-launchd-restart-resilience-design.md:95]: 💡 **SUGGESTION: Document the systemd quoting behavior**

The design doc mentions that paths are "double-quoted" but doesn't explain the systemd-specific requirements. Consider adding a note about the quoting strategy:

```markdown
### Path Quoting

Paths in the systemd unit file use single quotes (`'...'`) rather than double quotes 
to ensure proper handling of paths containing spaces. systemd strips one layer of 
outer quotes from ExecStart and WorkingDirectory values, so:
- `ExecStart='/path with spaces/node'` works correctly
- `ExecStart="/path with spaces/node"` would fail (quotes treated literally)
```

This helps future maintainers understand why the implementation differs from typical shell quoting.

File: tests/service-manager.test.cjs [hunk 8]
```diff
@@ -0,0 +1,209 @@
+/**
```
Comment (Tara-ag) [inline tests/service-manager.test.cjs:1]: 💡 **SUGGESTION: Add test for empty/invalid path inputs**

Consider adding defensive tests for edge cases in the builder functions:

```javascript
runTest('buildLaunchdPlist: handles empty strings gracefully', () => {
  const plist = buildLaunchdPlist({
    label: '',
    nodeBin: '',
    shooterBin: '',
    pkgRoot: '',
    pathEnv: '',
    shooterHome: '',
    logFile: '',
  });
  // Should still produce valid XML structure
  assertTrue(plist.startsWith('<?xml'), 'produces valid XML even with empty inputs');
  assertTrue(plist.endsWith('</plist>'), 'XML is properly closed');
});
```

This documents the expected behavior when paths resolve unexpectedly and ensures the plist/unit generation doesn't crash on edge cases.

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Review Summary

**Files reviewed:** 5 files changed (+602/-63 lines)

### Issues Found

| Severity | Count | Description |
|----------|-------|-------------|
| 🔒 CRITICAL | 0 | None |
| ⚠️ MAJOR | 1 | Systemd unit file quoting for paths with spaces |
| 💡 MINOR | 2 | ppid===1 reliability, test coverage gaps |
| 💬 SUGGESTION | 5 | Error handling, documentation, race conditions |

### Blocking Issue (MAJOR)

**`bin/lib/service-manager.cjs:118` - Incorrect systemd quoting**

The systemd unit file uses `"` for path quoting which systemd treats as literal characters, not shell-style quotes. This will cause service startup failures when node or shooter are installed in paths containing spaces (e.g., `/Users/John Doe/...`).

**Fix required:** Use single quotes (`'...'`) instead of double quotes for `ExecStart` and `WorkingDirectory` values in the systemd unit template.

### Non-blocking Observations

1. **Transition fallback reliability**: The `ppid === 1` check for legacy plist detection is a reasonable heuristic but should be documented as temporary/best-effort.

2. **Test coverage**: Good coverage of the core decider functions. Consider adding tests for paths with spaces once the quoting fix is applied.

3. **Design**: The service-manager-owned lifecycle approach is idiomatic and correctly solves the restart-resilience problem described in the PR.

### Next Steps

1. Fix the systemd quoting issue in `buildSystemdUnit()`
2. Add/update tests for paths with spaces
3. Consider documenting the ppid fallback as transitional

Once the systemd quoting is corrected, this PR should be ready for approval.

File: bin/lib/service-manager.cjs (hunk 1 above)
Comment (murdore) [reply bin/lib/service-manager.cjs:123]: The `ExecStart` quoting here is unchanged from the pre-PR `enableSystemdUnit` (extracted verbatim into the builder), so it isn't a regression introduced by this change. For `ExecStart=`, systemd strips boundary quotes (`EXTRACT_UNQUOTE`), so `ExecStart="/path with spaces/node" "…" start` parses as two whole arguments — double quotes at a token boundary are valid per systemd.syntax(7). The interpolated values (node path, shooter bin, `$HOME`) don't normally contain spaces on Linux, so the quoting is defensive. I've added a test (`buildSystemdUnit: double-quotes ExecStart tokens …`) to lock that contract. Since systemd/Linux is the rarely-used path and I can't validate runtime parsing changes without a systemd host, I'd rather harden quoting (incl. `WorkingDirectory=`) in a dedicated, Linux-tested change than alter it untested here.

File: bin/shooter.cjs (hunk 2 above)
Comment (murdore) [reply bin/shooter.cjs:1161]: Expanded the comment: `ppid === 1` is now called out as a best-effort heuristic that won't hold under sudo/wrappers, exists only to avoid delegate-recursion for old marker-less plists, and can be dropped once all installs carry the `SHOOTER_MANAGED` marker.

File: tests/service-manager.test.cjs (hunk 3 above)
Comment (murdore) [reply tests/service-manager.test.cjs:211]: Added `buildSystemdUnit: double-quotes ExecStart tokens so spaced paths stay one arg`, asserting both the double-quoted `ExecStart` tokens and the double-quoted `Environment=SHOOTER_HOME` value for a path containing spaces.

File: bin/shooter.cjs (hunk 4 above)
Comment (murdore) [reply bin/shooter.cjs:1146]: Returning `false` when `systemctl` is absent is the intended result — no systemd means not systemd-managed. The silent catch is deliberately symmetric with the existing `isLaunchdManaging()`, which swallows its `launchctl print` failure the same way. There's no debug-log facility for these probes in the CLI, so adding one here (and not to the launchd twin) would be inconsistent. Leaving as-is.

File: tests/service-manager.test.cjs (hunk 5 above)
Comment (murdore) [reply tests/service-manager.test.cjs:201]: Broadened the escape test to assert all of `&`, `<`, `>` are encoded — it now feeds `/p<a>&b` and checks for `<string>/p&lt;a&gt;&amp;b</string>`.

File: bin/shooter.cjs (hunk 6 above)
Comment (murdore) [reply bin/shooter.cjs:1179]: `shooter start` is a one-shot, human-invoked CLI command — nothing else calls `delegateStartToLaunchd` concurrently, so the check→bootstrap window can't actually race. If the job is already loaded, `isLaunchdManaging()` returns early; a failed `bootstrap` is caught and surfaced. Adding locking for a path that can't be hit would be complexity without benefit, so I'll leave it.

File: docs/superpowers/specs/2026-06-27-launchd-restart-resilience-design.md (hunk 7 above)
Comment (murdore) [reply docs/superpowers/specs/2026-06-27-launchd-restart-resilience-design.md:95]: Tied to the `ExecStart`-quoting thread: the quoting is preserved from the original implementation and valid for `ExecStart`, so I'm keeping the design doc scoped to the restart-resilience change rather than documenting pre-existing systemd quoting. The rationale lives in that thread and the new `buildSystemdUnit` spaces test.

File: tests/service-manager.test.cjs (hunk 8 above)
Comment (murdore) [reply tests/service-manager.test.cjs:1]: The builder inputs are always populated system values (`process.execPath`, `os.homedir()`/`SHOOTER_HOME`, `PKG_ROOT`, the fixed label) — no call site passes empty strings, so a test for that state would assert behavior that can't occur. Skipping it (YAGNI) to keep the suite on real inputs.

**Comment by murdore** [comment]
## Review feedback addressed

Squashed to a single commit (Single Commit Policy passes). Replied in-thread to all 8 comments.

**Changed (3):**
- ppid comment expanded to mark it a best-effort, removable heuristic.
- Broadened the launchd XML-escape test to cover `&`, `<`, `>`.
- Added a `buildSystemdUnit` test locking the double-quoted `ExecStart`/`Environment` contract for paths with spaces.

**Pushed back, with reasoning in-thread (5):**
- **systemd quoting (MAJOR):** unchanged from the pre-PR `enableSystemdUnit`; double quotes at a token boundary are valid `ExecStart` syntax (systemd strips them via `EXTRACT_UNQUOTE`). Hardening `WorkingDirectory=` belongs in a dedicated, Linux-tested change.
- **systemctl silent-catch:** returning `false` when systemd is absent is correct and symmetric with the existing `isLaunchdManaging()`.
- **TOCTOU in delegateStartToLaunchd:** `shooter start` is a one-shot CLI command with no concurrent caller — the window can't race.
- **doc / empty-path tests:** scope + YAGNI (inputs are always populated system paths).

Tests: 14/14 in `service-manager.test.cjs`, full `pnpm test` green locally.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Files reviewed:** 5 (all changed files)
**New issues raised:** 0
**Existing review threads:** 8 (all addressed by author)

### Assessment

This PR fixes a legitimate operational issue where clean SIGTERM exits were not being restarted by launchd due to `KeepAlive = { SuccessfulExit: false }`. The solution follows idiomatic service-manager patterns:

- **Correctness:** The delegation logic with `SHOOTER_MANAGED` marker prevents recursion, and the `ppid === 1` fallback (documented as transitional) provides backward compatibility.

- **Security:** No concerns identified. All command executions use `execFileSync` with array arguments (no shell injection). Paths are constructed from system values (`os.homedir()`, `process.execPath`) + constants, not user input. XML escaping is applied via the existing `xmlEscapeText` utility.

- **Testing:** 13 new unit tests cover the pure decider functions and unit-file generation, added to the test suite in `package.json`.

- **Documentation:** Comprehensive design doc explains the problem, solution, and deployment notes.

### Existing Review Threads

All 8 review comments from Tara-ag have been appropriately addressed:
- Systemd quoting preserved from pre-PR (valid per systemd.syntax(7))
- `ppid === 1` heuristic documented as best-effort transition fallback
- Test coverage expanded for XML escaping (`&`, `<`, `>`) and path quoting
- Error handling for missing `systemctl` deemed intentional (symmetric with `isLaunchdManaging()`)

**Recommendation:** Approve. The fix addresses the root cause (17h outage from clean exit not restarting) with a robust service-manager-owned lifecycle approach.

File: bin/shooter.cjs [hunk 9]
```diff
@@ -315,6 +321,18 @@ function startServer() {
...
+  // service manager) hands off to launchd/systemd so the OS owns restart and
+  // recovers from any clean SIGTERM. The manager-spawned instance carries
+  // SHOOTER_MANAGED=1 (or ppid 1 on macOS) and falls through to run directly.
+  const startAction = resolveStartAction({
+    managed: isSpawnedByManager(),
+    agentInstalled: isAutostartInstalled(),
+    platform: os.platform(),
+  });
+  if (startAction === 'delegate-launchd') return delegateStartToLaunchd();
+  if (startAction === 'delegate-systemd') return delegateStartToSystemd();
```
[the remark this answers is not in the document]
Comment (murdore) [reply bin/shooter.cjs:340]: Fixed — `startServer` now computes `hasRuntimeOverrides` (`--port`/`-p`, `--daemon`/`-d`, `--no-tunnel`) and passes it to `resolveStartAction`, which returns `run-server` when set. So `shooter start --port 9000` (or `--daemon`/`--no-tunnel`) on an autostart machine runs directly and honours the flags instead of delegating to the static unit. Covered by a new test (`resolveStartAction: runtime overrides … run directly`).

File: bin/shooter.cjs [hunk 10]
```diff
@@ -1153,6 +1140,99 @@ function isLaunchdManaging() {
...
+  }
+  try {
+    execFileSync('launchctl', ['bootstrap', `gui/${uid}`, LAUNCHD_PLIST], { stdio: 'ignore' });
+  } catch (err) {
+    console.error(`Failed to start via launchd: ${err.message}`);
+    console.error('Try: shooter autostart on');
+    process.exit(1);
+  }
+  console.log('Shooter started (launchd-managed).');
+  console.log('  Logs:   shooter logs');
```
[the remark this answers is not in the document]
Comment (murdore) [reply bin/shooter.cjs]: Fixed — `delegateStartToLaunchd` now bootstraps when the job is absent and `kickstart -k`s when it's already loaded, matching the design doc's `bootstrap if absent, else kickstart -k` contract (and giving `shooter start` a way to recycle a loaded-but-stale agent). Applied the same shape to `delegateStartToSystemd` (`restart` when active, else `start`).

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Files reviewed:** 5 (all changed files)
**New issues found:** 0 blocking, 0 major, 2 minor suggestions
**Existing comments:** 10 threads (all resolved)

### Analysis

This PR fixes a critical production issue where the Shooter daemon would not restart after a clean SIGTERM due to `KeepAlive = { SuccessfulExit: false }` in the LaunchAgent plist. The solution is well-architected:

**Security:** ✅ No issues
- All `execFileSync` calls use array arguments (no shell injection risk)
- `LAUNCHD_LABEL` is a hardcoded constant; UID comes from `process.getuid()`
- No user-controlled input reaches shell commands

**Correctness:** ✅ Solid
- `parsePortFlag()` correctly validates ports 0-65535 and handles `-p`/`--port` forms
- `hasRuntimeOverrides` properly detects runtime flags before delegation
- `isSpawnedByManager()` has proper two-tier detection with documented ppid===1 fallback
- The kickstart/bootstrap split for launchd and restart/start for systemd correctly handles stale agents

**Test Coverage:** ✅ Adequate
- 13 unit tests covering decision logic and unit-file generation
- Tests for XML escaping (`&`, `<`, `>`) and systemd path quoting with spaces
- Runtime override delegation logic is tested

### Minor Suggestions (non-blocking)

1. **Test gap - single quotes in systemd paths:** The systemd unit test covers spaces but not single quotes, which require `'\''` escaping in systemd. This is extremely rare in practice (usernames with apostrophes like "O'Brien"), and the current double-quoting will fail gracefully (service won't start, user sees error).

2. **Future cleanup:** Consider adding a TODO comment with a target version for removing the `ppid === 1` fallback once sufficient time has passed for users to regenerate their plists.

### Verification

- `pnpm test` passes (including new service-manager tests)
- Design doc accurately reflects the implementation
- No CLI backward compatibility issues — all existing commands/flags preserved
- No auth bypass or secret handling changes

**Decision:** Approve. The fix addresses a real production outage, the implementation is clean and well-tested, and all prior review feedback has been addressed.
